Add `editorFileExtension` when clause context

Krzysztof-Cieslak opened the pull request:
This adds file extension `when` clause context.

Some languages are using same languageId for multiple different types of files (with different extensions). It may be useful to let developers define command by file extension, and not only by languageId. For example in F#, we have single languageId (fsharp) for all different types of files (.fs - normal F# files, .fsx - F# script files, .fsi - F# signature files). We would like to be able to display some commands, or menu entries only for one of the types.

vbfox commented:
👍

jrieken commented:
Great we want that, a dupe is already on the October plan

jrieken commented on src/vs/editor/common/editorContextKeys.ts:
```diff
@@ -28,6 +28,7 @@ export namespace EditorContextKeys {
 
 	// -- mode context keys
 	export const languageId = new RawContextKey<string>('editorLangId', undefined);
+	export const fileExtension = new RawContextKey<string>('editorFileExtension', undefined);
```
Don't use `File` as it might not a file, maybe `editorResourceExtension` or `editorExtension`?

jrieken commented on src/vs/editor/common/modes/editorModeContext.ts:
```diff
@@ -111,6 +114,7 @@ export class EditorModeContext extends Disposable {
 			return;
 		}
 		this._langId.set(model.getLanguageIdentifier().language);
+		this._fileExtension.set(model.uri.path.split('.').pop());
```
😱 use `paths.extname`

Krzysztof-Cieslak commented on src/vs/workbench/common/resources.ts:
```diff
@@ -37,19 +39,22 @@ export class ResourceContextKey implements IContextKey<URI> {
+		this._extensionKey.set(value && paths.extname(value.fsPath));
 	}
 
 	reset(): void {
```
@jrieken. an reason why `_filenameKey` is not reseted here?

jrieken replied:
I'd say that is bug cc @bpasero

bpasero replied:
@jrieken introduced by you via https://github.com/Microsoft/vscode/commit/6110660a23471d10acdecd9f6834686576718411#diff-cb108a154667420fbda68f22a3d6d90c ?

jrieken replied:
🙉

Krzysztof-Cieslak replied:
Should I add this missing reset in this PR?

Krzysztof-Cieslak commented:
OK, I renamed context to `editorExtension` and added respective `resourceExtension`. Also used `paths` module in this one problematic place (didn't know before that you have those helpers)

jrieken commented:
re https://github.com/Microsoft/vscode/pull/34889#issuecomment-332567945

@Krzysztof-Cieslak Make sure to invoke "Format Document" before pushing. We run a formatter as part of hygiene and TypeScript formatting much match, whitespace rules must be happy, and the copyright statement must be correct

jrieken commented on src/vs/editor/common/editorContextKeys.ts:
```diff
@@ -28,6 +28,7 @@ export namespace EditorContextKeys {
 
 	// -- mode context keys
 	export const languageId = new RawContextKey<string>('editorLangId', undefined);
+	export const editorExtension = new RawContextKey<string>('editorExtension', undefined);
```
After a second look: we actually don't need to repeat this. Having `resourceExtname/Extension` is enough has it gets inherited to the editor from its column/slot. The reason for having `editorLangId` and `resourceLangId` is that latter is derived from the name while the former is more correct (look at the name and the contents)

jrieken commented on src/vs/workbench/common/resources.ts:
```diff
@@ -23,11 +23,13 @@ export class ResourceContextKey implements IContextKey<URI> {
 	static Filename = new RawContextKey<string>('resourceFilename', undefined);
 	static LangId = new RawContextKey<string>('resourceLangId', undefined);
 	static Resource = new RawContextKey<URI>('resource', undefined);
+	static Extension = new RawContextKey<string>('resourceExtension', undefined);
```
So, I'd prefer `resourceExtname`. What do you think?

jrieken commented on src/vs/workbench/common/resources.ts:
```diff
@@ -37,19 +39,22 @@ export class ResourceContextKey implements IContextKey<URI> {
 		this._schemeKey.set(value && value.scheme);
 		this._filenameKey.set(value && basename(value.fsPath));
 		this._langIdKey.set(value && this._modeService.getModeIdByFilenameOrFirstLine(value.fsPath));
+		this._extensionKey.set(value && paths.extname(value.fsPath));
```
Way to go but we to think if folks expect `.do` from `foo.do` or `do`...

Krzysztof-Cieslak replied:
I think typical behavior is to return extension with the dot. At least that's what most standard libraries I know do - Node's `path.extname`, .Net Framework `Path.GetExtension`, or Ruby's `File.extname`

Krzysztof-Cieslak replied:
More interesting question is if users expects it to be normalized to lower case (i.e if "ABC.FSX" should return ".FSX" or just ".fsx")

jrieken replied:
Good questions... I'd say we have it "as is". I can see how a "normalized" version can help but that applies to many values, e.g. filename

Krzysztof-Cieslak commented:
OK, removed new context from the ediotrContextKeys, leaving only resource one. And  I've renamed the resource on to `resourceExtname`

jrieken commented:
Thanks @Krzysztof-Cieslak
